### ISSUE TITLE
Debugging of Hand-Generated WASM

### ISSUE THREAD

**Issue by Dakror** [WebAssembly/binaryen#3462]
Hi there, for my usecase im generating WASM using the C API without coming from any other language to transpile from. But in order to debug the resulting code in a runtime such as wasmtime using GDB, as far as im understanding it, in the best case every expression has a debug information associated with it including file, line, col. 

Is there a better way to facilitate this for my use case than to go through all api calls and wrapping them in some logger to create the debug annotation? Is there an API detail im missing? 

Ideally i would just use the generated WAT file as a source file to use during step debugging.

**Comment by dcodeIO** [comment]
If all you are interested in is the WAT file, enabling debug info in binaries with `BinaryenSetDebugInfo` may help. When enabled, emitted binaries include a name section with function, global, local names etc. that engines read to actually name things instead of resorting to `$0` etc.

**Comment by Dakror** [comment]
Yeah i do that, but its still not useful for using step debugging for instance using a JIT compilation engine such as wasmtime. The only way they support proper debugging is by using the DWARF information, commonly emitted by compiler engines such as LLVM

**Comment by dcodeIO** [comment]
Hmm, yeah, a mechanism to more easily generate DWARF info is something I'd love to have as well. Not familiar with it yet, unfortunately, but perhaps someone else has an idea :)

**Comment by Dakror** [comment]
i've looked into libdwarf and man its a mess and also very complicated to add as an API user since you need all sorts of binary offsets and stuff.

**Comment by Dakror** [comment]
Also Binaryen already provides the functionality of recording those binary locations since DWARF information that exists will be updated according to transformations, so adding a function to just emit new DWARF information for the recorded WASM should be possible

**Comment by rennergade** [comment]
Did any of this ever get implemented? I'm having trouble using gdb on wasm binaries using asyncify.

**Comment by kripken** [comment]
Binaryen has good support for source maps, so using a debugger with that support (like a browser debugger, node.js debugger etc.) can be useful. Various toolchains use this, like Emscripten (C++), J2CL (Java), etc.

Binaryen has some partial DWARF support, but it gets worse with things like Asyncify because of all the code changes Asyncify does, so that is not recommended, unlike source maps.

Another option is to instrument the wasm, which does what the top comment here mentions,

> go through all api calls and wrapping them in some logger

The relevant passes are `--log-execution`, `--instrument-locals`, `--instrument-memory`.

**Comment by rennergade** [comment]
> Binaryen has good support for source maps, so using a debugger with that support (like a browser debugger, node.js debugger etc.) can be useful. Various toolchains use this, like Emscripten (C++), J2CL (Java), etc.
> 
> Binaryen has some partial DWARF support, but it gets worse with things like Asyncify because of all the code changes Asyncify does, so that is not recommended, unlike source maps.
> 
> Another option is to instrument the wasm, which does what the top comment here mentions,
> 
> > go through all api calls and wrapping them in some logger
> 
> The relevant passes are `--log-execution`, `--instrument-locals`, `--instrument-memory`.

Trying some of this out but maybe I'm confused. Are you saying that adding those flags should make a wasm program with asyncify compatible with gdb? If so would I use all of those flags together or on separate runs of wasm-opt?

Thanks!

**Comment by kripken** [comment]
Sorry for being unclear, those flags are another way to debug, manually, that does not use DWARF.

If you want to use DWARF (and not source maps or manual debugging with those passes) then you can try it (with the limitations mentioned above). But I don't know if gdb has support for wasm files. What I do know works is [DWARF in the Chrome devtools](https://developer.chrome.com/docs/devtools/wasm).

**Comment by rennergade** [comment]
> Sorry for being unclear, those flags are another way to debug, manually, that does not use DWARF.
> 
> If you want to use DWARF (and not source maps or manual debugging with those passes) then you can try it (with the limitations mentioned above). But I don't know if gdb has support for wasm files. What I do know works is [DWARF in the Chrome devtools](https://developer.chrome.com/docs/devtools/wasm).

Right, yeah. I can debug wasm programs via gdb that use DWARF but once I apply asyncify I lose that ability. It seems like there's not really an easy way around that.

**Comment by kripken** [comment]
`wasm-opt --asyncify` should not strip DWARF, though - it might lose some precision, but it should still be there. You can verify that by seeing that the DWARF sections (like `.debug_loc`) exist before and after running Asyncify. It's possible though that gdb's DWARF support if confused by them, as this has mostly been tested against browser devtools (so it might be worth trying there, with the link from earlier).

**Comment by tlively** [comment]
Closing this because there does not seem to be anything actionable here. If there is a specific bug with a reproducer, please open a new issue.
